client/fetch_tagged_results_accumulator.go

```diff
@@ -0,0 +1,366 @@
+	sort.Sort(results)
+	accum.responses = fetchTaggedIDResults(results)
+
+	numElements := 0
+	accum.responses.forEachID(func(_ fetchTaggedIDResults) bool {
+		numElements++
+		return numElements < limit
+	})
+
+	result := pools.MutableSeriesIterators().Get(numElements)
+	result.Reset(numElements)
+	count := 0
+	accum.responses.forEachID(func(elems fetchTaggedIDResults) bool {
+		seriesIter := accum.sliceResponsesAsSeriesIter(pools, elems)
+		result.SetAt(count, seriesIter)
+		count++
+		return count < limit
+	})
+
+	exhaustive := accum.exhaustive && count < limit
+	return result, exhaustive, nil
+}
+
+func (accum *fetchTaggedResultAccumulator) AsIndexQueryResults(
+	limit int,
+	pools fetchTaggedPools,
+) (index.QueryResults, error) {
+	var (
+		iter  = newFetchTaggedResultsIndexIterator(pools)
+		count = 0
+	)
+	results := fetchTaggedIDResultsSortedByID(accum.responses)
+	sort.Sort(results)
+	accum.responses = fetchTaggedIDResults(results)
+	accum.responses.forEachID(func(elems fetchTaggedIDResults) bool {
+		iter.backing.ids = append(iter.backing.ids, elems[0].ID)
+		iter.backing.nses = append(iter.backing.nses, elems[0].NameSpace)
+		iter.backing.tags = append(iter.backing.tags, elems[0].EncodedTags)
+		count++
+		return count < limit
+	})
+
+	return index.QueryResults{
+		Exhaustive: accum.exhaustive && count < limit,
+		Iterator:   iter,
+	}, nil
+}
+
+type fetchTaggedShardConsistencyResults []fetchTaggedShardConsistencyResult
+
+func (res fetchTaggedShardConsistencyResults) initialize(length int) fetchTaggedShardConsistencyResults {
+	if cap(res) < length {
+		res = append(res, make(fetchTaggedShardConsistencyResults, length-cap(res))...)
+	}
+	res = res[:length]
+	// following compiler optimized memcpy impl:
+	// https://github.com/golang/go/wiki/CompilerOptimizations#optimized-memclr
+	for i := range res {
+		res[i] = fetchTaggedShardConsistencyResult{}
+	}
+	return res
+}
+
+type fetchTaggedIDResults []*rpc.FetchTaggedIDResult_
+
+// lambda to iterate over fetchTagged responses a single id at a time,
+// the returned bool indicates if the iteration should be continued past the
+// curent id.
+type forEachFetchTaggedIDFn func(responsesForSingleID fetchTaggedIDResults) bool
+
+// forEachID iterates over the provide results, and calls `fn` on each
+// group of responses with the same ID.
+// NB: assumes the results array being operated upon has been sorted
+func (results fetchTaggedIDResults) forEachID(fn forEachFetchTaggedIDFn) {
+	var (
+		startIdx = 0
+		lastID   []byte
+	)
+	for i := 0; i < len(results); i++ {
+		elem := results[i]
+		if !bytes.Equal(elem.ID, lastID) {
+			// We only want to call the the forEachID fn once we have calculated the entire group,
+			// i.e. once we have gone past the last element for a given ID, but the first element
+			// in the results slice is a special case because we are always starting a new group
+			// at that point.
+			if i == 0 {
+				lastID = elem.ID
```

Comment:
Perhaps place `lastID = elem.ID` as first line of inside `if !bytes.Equal(...) { ... }` clause?  that way don't have to duplicate it twice.

Reply:
Sure thing.